File: src/mysqlcookbook/recipes/api/06_conn_params/cmdline.rb
#!/usr/bin/ruby
# cmdline.rb: demonstrate command-line option parsing in Ruby

require "getoptlong"
require "mysql2"

# connection parameters - all missing (nil) by default
host_name = nil
password = nil
user_name = nil

opts = GetoptLong.new(
  [ "--host",     "-h", GetoptLong::REQUIRED_ARGUMENT ],
  [ "--password", "-p", GetoptLong::REQUIRED_ARGUMENT ],
  [ "--user",     "-u", GetoptLong::REQUIRED_ARGUMENT ]
)

# iterate through options, extracting whatever values are present;
# opt is the long-format option, arg is its value
opts.each do |opt, arg|
  case opt
  when "--host"
    host_name = arg
  when "--password"
    password = arg
  when "--user"
    user_name = arg
  end
end

# any nonoption arguments remain in ARGV
# and can be processed here as necessary

# connect to server
begin
  if !host_name.nil?
    client = Mysql2::Client.new(:host => host_name, 
                                :username => user_name, 
                                :password => password,
                                :databse => "cookbook")
  else
    client = Mysql2::Client.new(:username => user_name, 
                                :password => password,
                                :databse => "cookbook")
  end
  puts "Connected"
rescue Mysql2::Error => e
  puts "Cannot connect to server"
  puts "Error code: #{e.errno}"
  puts "Error message: #{e.message}"
  exit(1)
end

client.close
puts "Disconnected"

File: src/mysqlcookbook/recipes/api/06_conn_params/cmdline.py
#!/usr/bin/python3
# cmdline.py: demonstrate command-line option parsing in Python

import sys
import getopt
import mysql.connector

try:
  opts, args = getopt.getopt(sys.argv[1:],
                             "h:p:u:",
                             [
                               "host=",
                                "password=",
                                "user="
                             ])
except getopt.error as e:
  # for errors, print program name and text of error message
  print("%s: %s" % (sys.argv[0], e))
  sys.exit(1)

# default connection parameter values
conn_params = {
  "database": "cookbook",
  "host": "",
  "user": "",
  "password": ""
}

# iterate through options, extracting whatever values are present
for opt, arg in opts:
  if opt in ("-h", "--host"):
    conn_params["host"] = arg
  elif opt in ("-p", "--password"):
    conn_params["password"] = arg
  elif opt in ("-u", "--user"):
    conn_params["user"] = arg

# any nonoption arguments remain in args
# and can be processed here as necessary

try:
  conn = mysql.connector.connect(**conn_params)
  print("Connected")
except mysql.connector.Error as e:
  print("Cannot connect to server")
  print("Error code: %s" % e.errno)
  print("Error message: %s" % e.msg)
else:
  conn.close()
  print("Disconnected")
